src/rootfs/udevd/pkg/drivers/scsi/scsi.go

```diff
@@ -0,0 +1,392 @@
+	ptr := uintptr(unsafe.Pointer(&version))
+
+	var f *os.File
+	if f, err = open(name); err != nil {
+		return nil, err
+	}
+
+	if err = ioctl(f.Fd(), SgGetVersionNum, ptr); err != nil {
+		return nil, err
+	}
+
+	if version < 30000 {
+		return nil, errors.Errorf("no sg driver found for device at %s", name)
+	}
+
+	device = &Device{
+		f: f,
+		ID: &ID{
+			UseSg: 3,
+		},
+	}
+
+	return device, nil
+}
+
+// Close closes the device.
+func (dvc *Device) Close() error {
+	return dvc.f.Close()
+}
+
+// Identify queries a SCSI device via the SCSI INQUIRY vital product data (VPD)
+// page 0x80 and 0x83.
+func (dvc *Device) Identify() (err error) {
+	// Standard inquiry.
+	if err = dvc.StandardInquiry(); err != nil {
+		return errors.Errorf("standard inquiry failed: %v", err)
+	}
+
+	// Page 0 inquiry.
+	var pages []byte
+	if pages, err = dvc.Page0Inquiry(); err != nil {
+		return errors.Errorf("page 0 inquiry failed: %v", err)
+	}
+	// TODO(andrewrynhard): Prevent Page80Inquiry from running twice.
+	for _, page := range pages {
+		if page == Page80 {
+			// Page 80 inquiry.
+			if err = dvc.Page80Inquiry(); err != nil {
+				return errors.Errorf("page 80 inquiry failed: %v", err)
+			}
+		}
+		if page == Page83 {
+			// Page 80 inquiry.
+			if err = dvc.Page80Inquiry(); err != nil {
+				return errors.Errorf("page 80 inquiry failed: %v", err)
+			}
+			// Page 83 inquiry.
+			if err = dvc.Page83Inquiry(); err != nil {
+				return errors.Errorf("page 83 inquiry failed: %v", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// StandardInquiry performs a basic inquiry on the device.
+// nolint: gocyclo
+func (dvc *Device) StandardInquiry() (err error) {
+	var info os.FileInfo
+	if info, err = dvc.f.Stat(); err != nil {
+		return errors.Errorf("stat failed: %v", err)
+	}
+
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return errors.New("not a syscall.Stat_t")
+	}
+
+	dvc.ID.Kernel = fmt.Sprintf("%d:%d", unix.Major(stat.Rdev), unix.Minor(stat.Rdev))
+
+	var buf CommandDescriptorBlock
+	if buf, err = dvc.inquiry(0, 0); err != nil {
+		return err
+	}
+
+	dvc.Vendor = string(buf[8:16])
+	dvc.Model = string(buf[16:32])
+	dvc.Revision = string(buf[32:36])
+
+	t := buf[1] & 0x1f
+	switch t {
+	case 0:
+		dvc.Type = TypeDisk
+	case 1:
+		dvc.Type = TypeTape
+	case 4:
+		dvc.Type = TypeOptical
+	case 5:
+		dvc.Type = TypeCD
+	case 7:
+		dvc.Type = TypeOptical
+	case 0xe:
+		dvc.Type = TypeDisk
+	case 0xf:
+		dvc.Type = TypeOptical
+	default:
+	}
+
+	return nil
+}
+
+// Page0Inquiry returns a list of Vital Product Data (VPD) pages.
+// TODO(andrewrynhard): Should pages by an iota?
+func (dvc *Device) Page0Inquiry() (pages []byte, err error) {
+	var buf CommandDescriptorBlock
+
+	if buf, err = dvc.inquiry(1, Page0); err != nil {
+		return nil, err
+	}
+
+	if buf[3] > InguiryBufLen {
+		return nil, errors.Errorf("%s: page 0 buffer too big", dvc.Kernel)
+	}
+
+	pages = []byte{}
+	for i := 4; i < int(buf[3])+3; i++ {
+		switch buf[i] {
+		case Page80:
+			fallthrough
```

Comment:
does this work? it should just cause it to try the next case statement which would fail since Page80 != Page83 right?